OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,406 @@
+        /// <param name="odpEvent"></param>
+        private void Enqueue(OdpEvent odpEvent)
+        {
+            if (_queue.Count >= _queueSize)
+            {
+                _logger.Log(LogLevel.WARN, $"ODP event send failed (queueSize = {_queue.Count}).");
+                return;
+            }
+
+            _queue.Enqueue(odpEvent);
+
+            ProcessQueue();
+        }
+
+        /// <summary>
+        /// Process the queue
+        /// </summary>
+        private void ProcessQueue()
+        {
+            if (State != ExecutionState.Running)
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG, $"Processing Queue.");
+
+            State = ExecutionState.Processing;
+            while (QueueHasBatches())
+            {
+                DequeueSendSingleBatch();
+            }
+
+            State = ExecutionState.Running;
+        }
+
+        private void RegularlyFlushQueue()
+        {
+            while (!_flushIntervalCancellation.IsCancellationRequested)
+            {
+                Task.Delay(_flushInterval).ContinueWith(_ => FlushQueue()).Wait();
+            }
+        }
+
+        private void FlushQueue()
+        {
+            if (State != ExecutionState.Running)
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG, $"Flushing Queue.");
+
+            State = ExecutionState.Processing;
+            while (QueueContainsItems())
+            {
+                DequeueSendSingleBatch();
+            }
+
+            State = ExecutionState.Running;
+        }
+
+        /// <summary>
+        /// Dequeue a single batch and send it
+        /// </summary>
+        private void DequeueSendSingleBatch()
+        {
+            var batch = new List<OdpEvent>(_batchSize);
+
+            // dequeue a batch 
+            for (int i = 0; i < _batchSize && _queue.Count > 0; i += 1)
+            {
+                if (_queue.TryDequeue(out OdpEvent dequeuedOdpEvent))
+                {
+                    batch.Add(dequeuedOdpEvent);
+                }
+            }
+
+            if (batch.Count > 0)
+            {
+                Task.Run(() =>
+                {
+                    bool shouldRetry;
+                    var attemptNumber = 0;
+                    do
+                    {
+                        shouldRetry = _odpEventApiManager.SendEvents(_odpConfig.ApiKey,
+                            _odpConfig.ApiHost, batch);
+                        attemptNumber += 1;
+                    } while (shouldRetry && attemptNumber < MAX_RETRIES);
+                });
+            }
+        }
+
+        /// <summary>
+        /// Determines if the ODP configuration is ready
+        /// </summary>
+        /// <returns>True if required parameters are available otherwise False</returns>
+        private bool IsOdpConfigurationReady()
+        {
+            if (_odpConfig.IsReady())
+            {
+                return true;
```

Comment:
can't we directly check from _odpConfig?

Reply:
The additional functionality of the not-ready config is needed so I'm encapsulating into this method.